File: android/app/src/main/java/com/mobilelivecaster/streaming/MediaTimestampTracker.kt
package com.mobilelivecaster.streaming

import kotlin.math.abs

internal class MediaTimestampTracker(
    private val warningThresholdMs: Long = 150L,
    private val criticalThresholdMs: Long = 500L,
    private val criticalConsecutiveSamples: Long = 3L,
    private val monotonicClockMs: () -> Long = { System.nanoTime() / 1_000_000L }
) {
    private var generationEpochObservedAtMs: Long? = null
    private var firstVideoTimestampMs: Long? = null
    private var firstAudioTimestampMs: Long? = null
    private var firstVideoObservedAtMs: Long? = null
    private var firstAudioObservedAtMs: Long? = null
    private var latestVideoTimestampMs = 0L
    private var latestAudioTimestampMs = 0L
    private var videoRevision = 0L
    private var audioRevision = 0L
    private var lastSampleVideoRevision = 0L
    private var lastSampleAudioRevision = 0L
    private var sampleCount = 0L
    private var outOfSyncSampleCount = 0L
    private var outOfSyncIncidentCount = 0L
    private var criticalIncidentCount = 0L
    private var consecutiveOutOfSyncSamples = 0L
    private var maxConsecutiveOutOfSyncSamples = 0L
    private var consecutiveCriticalSamples = 0L
    private var maxAbsSkewMs = 0L
    private var wasOutOfSync = false
    private var wasCritical = false

    init {
        require(warningThresholdMs > 0L) { "A/V sync warning threshold must be positive" }
        require(criticalThresholdMs >= warningThresholdMs) { "A/V sync critical threshold must not be below warning" }
        require(criticalConsecutiveSamples > 0L) { "A/V sync critical sample count must be positive" }
    }

    @Synchronized
    fun reset() {
        generationEpochObservedAtMs = null
        firstVideoTimestampMs = null
        firstAudioTimestampMs = null
        firstVideoObservedAtMs = null
        firstAudioObservedAtMs = null
        latestVideoTimestampMs = 0L
        latestAudioTimestampMs = 0L
        videoRevision = 0L
        audioRevision = 0L
        lastSampleVideoRevision = 0L
        lastSampleAudioRevision = 0L
        sampleCount = 0L
        outOfSyncSampleCount = 0L
        outOfSyncIncidentCount = 0L
        criticalIncidentCount = 0L
        consecutiveOutOfSyncSamples = 0L
        maxConsecutiveOutOfSyncSamples = 0L
        consecutiveCriticalSamples = 0L
        maxAbsSkewMs = 0L
        wasOutOfSync = false
        wasCritical = false
    }

    @Synchronized
    fun startGeneration() {
        generationEpochObservedAtMs = null
        firstVideoTimestampMs = null
        firstAudioTimestampMs = null
        firstVideoObservedAtMs = null
        firstAudioObservedAtMs = null
        latestVideoTimestampMs = 0L
        latestAudioTimestampMs = 0L
        videoRevision = 0L
        audioRevision = 0L
        lastSampleVideoRevision = 0L
        lastSampleAudioRevision = 0L
        consecutiveOutOfSyncSamples = 0L
        consecutiveCriticalSamples = 0L
        wasOutOfSync = false
        wasCritical = false
    }

    @Synchronized
    fun recordVideo(timestampMs: Long, observedAtMs: Long = monotonicClockMs()): NativeRuntimeAvSync {
        val normalized = timestampMs.coerceAtLeast(0L)
        val observed = observedAtMs.coerceAtLeast(0L)
        val epoch = generationEpochObservedAtMs ?: observed.also { generationEpochObservedAtMs = it }
        val baseline = firstVideoTimestampMs ?: normalized.also { firstVideoTimestampMs = it }
        val observedBaseline = firstVideoObservedAtMs ?: observed.also { firstVideoObservedAtMs = it }
        latestVideoTimestampMs = ((observedBaseline - epoch) + (normalized - baseline)).coerceAtLeast(latestVideoTimestampMs)
        videoRevision += 1L
        recordSampleIfReady()
        return snapshot()
    }

    @Synchronized
    fun recordAudio(timestampMs: Long, observedAtMs: Long = monotonicClockMs()): NativeRuntimeAvSync {
        val normalized = timestampMs.coerceAtLeast(0L)
        val observed = observedAtMs.coerceAtLeast(0L)
        val epoch = generationEpochObservedAtMs ?: observed.also { generationEpochObservedAtMs = it }
        val baseline = firstAudioTimestampMs ?: normalized.also { firstAudioTimestampMs = it }
        val observedBaseline = firstAudioObservedAtMs ?: observed.also { firstAudioObservedAtMs = it }
        latestAudioTimestampMs = ((observedBaseline - epoch) + (normalized - baseline)).coerceAtLeast(latestAudioTimestampMs)
        audioRevision += 1L
        recordSampleIfReady()
        return snapshot()
    }

    @Synchronized
    fun snapshot(): NativeRuntimeAvSync {
        val ready = firstVideoTimestampMs != null && firstAudioTimestampMs != null
        val skewMs = if (ready) latestVideoTimestampMs - latestAudioTimestampMs else 0L
        val absSkewMs = abs(skewMs)
        val critical = ready && consecutiveCriticalSamples >= criticalConsecutiveSamples
        val confirmedOutOfSync = consecutiveOutOfSyncSamples >= criticalConsecutiveSamples
        val status = when {
            !ready -> "warming-up"
            absSkewMs <= warningThresholdMs || !confirmedOutOfSync -> "in-sync"
            skewMs > 0L -> "video-leading"
            else -> "audio-leading"
        }
        return NativeRuntimeAvSync(
            status = status,
            latestVideoTimestampMs = latestVideoTimestampMs,
            latestAudioTimestampMs = latestAudioTimestampMs,
            skewMs = skewMs,
            maxAbsSkewMs = maxAbsSkewMs,
            sampleCount = sampleCount,
            outOfSyncSampleCount = outOfSyncSampleCount,
            outOfSyncIncidentCount = outOfSyncIncidentCount,
            criticalIncidentCount = criticalIncidentCount,
            consecutiveOutOfSyncSamples = consecutiveOutOfSyncSamples,
            maxConsecutiveOutOfSyncSamples = maxConsecutiveOutOfSyncSamples,
            warningThresholdMs = warningThresholdMs,
            criticalThresholdMs = criticalThresholdMs,
            critical = critical
        )
    }

    private fun recordSampleIfReady() {
        if (firstVideoTimestampMs == null || firstAudioTimestampMs == null) return
        if (videoRevision <= lastSampleVideoRevision || audioRevision <= lastSampleAudioRevision) return
        lastSampleVideoRevision = videoRevision
        lastSampleAudioRevision = audioRevision
        val absSkewMs = abs(latestVideoTimestampMs - latestAudioTimestampMs)
        sampleCount += 1L
        maxAbsSkewMs = maxOf(maxAbsSkewMs, absSkewMs)
        val outOfSync = absSkewMs > warningThresholdMs
        if (outOfSync) {
            outOfSyncSampleCount += 1L
            consecutiveOutOfSyncSamples += 1L
            maxConsecutiveOutOfSyncSamples = maxOf(maxConsecutiveOutOfSyncSamples, consecutiveOutOfSyncSamples)
        } else {
            consecutiveOutOfSyncSamples = 0L
        }
        val confirmedOutOfSync = outOfSync && consecutiveOutOfSyncSamples >= criticalConsecutiveSamples
        if (absSkewMs >= criticalThresholdMs) {
            consecutiveCriticalSamples += 1L
        } else {
            consecutiveCriticalSamples = 0L
        }
        val critical = consecutiveCriticalSamples >= criticalConsecutiveSamples
        if (confirmedOutOfSync && !wasOutOfSync) outOfSyncIncidentCount += 1L
        if (critical && !wasCritical) criticalIncidentCount += 1L
        wasOutOfSync = confirmedOutOfSync
        wasCritical = critical
    }
}

internal class NativeRuntimeAvSyncAccumulator {
    private var retained: NativeRuntimeAvSync? = null

    @Synchronized
    fun reset() {
        retained = null
    }

    @Synchronized
    fun retain(snapshot: NativeRuntimeAvSync) {
        retained = merge(retained, snapshot)
    }

    @Synchronized
    fun combine(current: NativeRuntimeAvSync?): NativeRuntimeAvSync? = merge(retained, current)

    private fun merge(
        previous: NativeRuntimeAvSync?,
        current: NativeRuntimeAvSync?
    ): NativeRuntimeAvSync? {
        if (previous == null) return current
        if (current == null) return previous
        return current.copy(
            maxAbsSkewMs = maxOf(previous.maxAbsSkewMs, current.maxAbsSkewMs),
            sampleCount = previous.sampleCount + current.sampleCount,
            outOfSyncSampleCount = previous.outOfSyncSampleCount + current.outOfSyncSampleCount,
            outOfSyncIncidentCount = previous.outOfSyncIncidentCount + current.outOfSyncIncidentCount,
            criticalIncidentCount = previous.criticalIncidentCount + current.criticalIncidentCount,
            maxConsecutiveOutOfSyncSamples = maxOf(
                previous.maxConsecutiveOutOfSyncSamples,
                current.maxConsecutiveOutOfSyncSamples
            )
        )
    }
}

File: ios/MobileLiveCasterBroadcastUpload/MediaTimestampTracker.swift
import Foundation

struct BroadcastMediaTimestampSnapshot: Equatable {
    let status: String
    let latestVideoTimestampMs: Int
    let latestAudioTimestampMs: Int
    let skewMs: Int
    let maxAbsSkewMs: Int
    let sampleCount: Int
    let outOfSyncSampleCount: Int
    let outOfSyncIncidentCount: Int
    let criticalIncidentCount: Int
    let consecutiveOutOfSyncSamples: Int
    let maxConsecutiveOutOfSyncSamples: Int
    let warningThresholdMs: Int
    let criticalThresholdMs: Int
    let critical: Bool

    func asDictionary() -> [String: Any] {
        [
            "status": status,
            "latestVideoTimestampMs": latestVideoTimestampMs,
            "latestAudioTimestampMs": latestAudioTimestampMs,
            "skewMs": skewMs,
            "maxAbsSkewMs": maxAbsSkewMs,
            "sampleCount": sampleCount,
            "outOfSyncSampleCount": outOfSyncSampleCount,
            "outOfSyncIncidentCount": outOfSyncIncidentCount,
            "criticalIncidentCount": criticalIncidentCount,
            "consecutiveOutOfSyncSamples": consecutiveOutOfSyncSamples,
            "maxConsecutiveOutOfSyncSamples": maxConsecutiveOutOfSyncSamples,
            "warningThresholdMs": warningThresholdMs,
            "criticalThresholdMs": criticalThresholdMs,
            "critical": critical
        ]
    }
}

struct BroadcastMediaTimestampTracker: Equatable {
    private let warningThresholdMs: Int
    private let criticalThresholdMs: Int
    private let criticalConsecutiveSamples: Int
    private var generationOriginTimestampMs: Int?
    private var hasVideoTimestamp = false
    private var hasAudioTimestamp = false
    private var latestVideoTimestampMs = 0
    private var latestAudioTimestampMs = 0
    private var videoRevision = 0
    private var audioRevision = 0
    private var lastSampleVideoRevision = 0
    private var lastSampleAudioRevision = 0
    private var sampleCount = 0
    private var outOfSyncSampleCount = 0
    private var outOfSyncIncidentCount = 0
    private var criticalIncidentCount = 0
    private var consecutiveOutOfSyncSamples = 0
    private var maxConsecutiveOutOfSyncSamples = 0
    private var consecutiveCriticalSamples = 0
    private var maxAbsSkewMs = 0
    private var wasOutOfSync = false
    private var wasCritical = false

    init(
        warningThresholdMs: Int = 150,
        criticalThresholdMs: Int = 500,
        criticalConsecutiveSamples: Int = 3
    ) {
        self.warningThresholdMs = max(1, warningThresholdMs)
        self.criticalThresholdMs = max(self.warningThresholdMs, criticalThresholdMs)
        self.criticalConsecutiveSamples = max(1, criticalConsecutiveSamples)
    }

    mutating func reset() {
        generationOriginTimestampMs = nil
        hasVideoTimestamp = false
        hasAudioTimestamp = false
        latestVideoTimestampMs = 0
        latestAudioTimestampMs = 0
        videoRevision = 0
        audioRevision = 0
        lastSampleVideoRevision = 0
        lastSampleAudioRevision = 0
        sampleCount = 0
        outOfSyncSampleCount = 0
        outOfSyncIncidentCount = 0
        criticalIncidentCount = 0
        consecutiveOutOfSyncSamples = 0
        maxConsecutiveOutOfSyncSamples = 0
        consecutiveCriticalSamples = 0
        maxAbsSkewMs = 0
        wasOutOfSync = false
        wasCritical = false
    }

    mutating func startGeneration() {
        generationOriginTimestampMs = nil
        hasVideoTimestamp = false
        hasAudioTimestamp = false
        latestVideoTimestampMs = 0
        latestAudioTimestampMs = 0
        videoRevision = 0
        audioRevision = 0
        lastSampleVideoRevision = 0
        lastSampleAudioRevision = 0
        consecutiveOutOfSyncSamples = 0
        consecutiveCriticalSamples = 0
        wasOutOfSync = false
        wasCritical = false
    }

    mutating func recordVideo(timestampMs: Int) -> BroadcastMediaTimestampSnapshot {
        let normalized = max(0, timestampMs)
        let origin = generationOriginTimestampMs ?? normalized
        if generationOriginTimestampMs == nil { generationOriginTimestampMs = normalized }
        hasVideoTimestamp = true
        latestVideoTimestampMs = max(latestVideoTimestampMs, normalized - origin)
        videoRevision += 1
        recordSampleIfReady()
        return snapshot()
    }

    mutating func recordAudio(timestampMs: Int) -> BroadcastMediaTimestampSnapshot {
        let normalized = max(0, timestampMs)
        let origin = generationOriginTimestampMs ?? normalized
        if generationOriginTimestampMs == nil { generationOriginTimestampMs = normalized }
        hasAudioTimestamp = true
        latestAudioTimestampMs = max(latestAudioTimestampMs, normalized - origin)
        audioRevision += 1
        recordSampleIfReady()
        return snapshot()
    }

    func snapshot() -> BroadcastMediaTimestampSnapshot {
        let ready = hasVideoTimestamp && hasAudioTimestamp
        let skewMs = ready ? latestVideoTimestampMs - latestAudioTimestampMs : 0
        let absoluteSkewMs = abs(skewMs)
        let critical = ready && consecutiveCriticalSamples >= criticalConsecutiveSamples
        let status: String
        let confirmedOutOfSync = consecutiveOutOfSyncSamples >= criticalConsecutiveSamples
        if !ready {
            status = "warming-up"
        } else if absoluteSkewMs <= warningThresholdMs || !confirmedOutOfSync {
            status = "in-sync"
        } else if skewMs > 0 {
            status = "video-leading"
        } else {
            status = "audio-leading"
        }
        return BroadcastMediaTimestampSnapshot(
            status: status,
            latestVideoTimestampMs: latestVideoTimestampMs,
            latestAudioTimestampMs: latestAudioTimestampMs,
            skewMs: skewMs,
            maxAbsSkewMs: maxAbsSkewMs,
            sampleCount: sampleCount,
            outOfSyncSampleCount: outOfSyncSampleCount,
            outOfSyncIncidentCount: outOfSyncIncidentCount,
            criticalIncidentCount: criticalIncidentCount,
            consecutiveOutOfSyncSamples: consecutiveOutOfSyncSamples,
            maxConsecutiveOutOfSyncSamples: maxConsecutiveOutOfSyncSamples,
            warningThresholdMs: warningThresholdMs,
            criticalThresholdMs: criticalThresholdMs,
            critical: critical
        )
    }

    private mutating func recordSampleIfReady() {
        guard hasVideoTimestamp, hasAudioTimestamp else { return }
        guard videoRevision > lastSampleVideoRevision, audioRevision > lastSampleAudioRevision else { return }
        lastSampleVideoRevision = videoRevision
        lastSampleAudioRevision = audioRevision
        let absoluteSkewMs = abs(latestVideoTimestampMs - latestAudioTimestampMs)
        sampleCount += 1
        maxAbsSkewMs = max(maxAbsSkewMs, absoluteSkewMs)
        let outOfSync = absoluteSkewMs > warningThresholdMs
        if outOfSync {
            outOfSyncSampleCount += 1
            consecutiveOutOfSyncSamples += 1
            maxConsecutiveOutOfSyncSamples = max(maxConsecutiveOutOfSyncSamples, consecutiveOutOfSyncSamples)
        } else {
            consecutiveOutOfSyncSamples = 0
        }
        let confirmedOutOfSync = outOfSync && consecutiveOutOfSyncSamples >= criticalConsecutiveSamples
        if absoluteSkewMs >= criticalThresholdMs {
            consecutiveCriticalSamples += 1
        } else {
            consecutiveCriticalSamples = 0
        }
        let critical = consecutiveCriticalSamples >= criticalConsecutiveSamples
        if confirmedOutOfSync && !wasOutOfSync { outOfSyncIncidentCount += 1 }
        if critical && !wasCritical { criticalIncidentCount += 1 }
        wasOutOfSync = confirmedOutOfSync
        wasCritical = critical
    }
}
